src/libsyntax/parse/parser.rs

```diff
@@ -3154,7 +3154,14 @@ impl<'a> Parser<'a> {
         // Parse: `for <src_pat> in <src_expr> <src_loop_block>`
 
         let pat = self.parse_pat()?;
-        self.expect_keyword(keywords::In)?;
+        if !self.eat_keyword(keywords::In) {
+            let in_span = self.prev_span.between(self.span);
+            let mut err = self.sess.span_diagnostic
+                .struct_span_err(in_span, "missing `in` in `for` loop");
+            err.span_label(in_span, "expected `in` here");
```

Comment:
Nit: The extra label here creates unnecessary duplication, IMO. The primary message already says that `in` is missing and the suggestion is already formatted as a label.